### PR TITLE
Yuchen/set tmp,two nodes

### PR DESCRIPTION
**Pull request by YuchenMu**
"set_tmp" new API, new tests and new documentation done.

### CODE REVIEW

**Comment by szaldivar** [comment]
I will wait for the two nodes PR to be merged so I do not repeat myself here

**Comment by YuchenMu** [comment]
> I will wait for the two nodes PR to be merged so I do not repeat myself here

I will update two nodes after the meeting

**Comment by szaldivar** [comment]
you might want to merge the changes you did in your other branch to fix the merge conflicts here

**Comment by YuchenMu** [comment]
Yes, that’s what I’m doing now. Thank you!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: src/models/set_tmp.js [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+  units_converter_array,
+  check_standard_compliance,
+  check_standard_compliance_array,
+  round,
+} from "../utilities/utilities.js";
+import { two_nodes, two_nodes_array, roundArray } from "../models/two_nodes.js";
+
+/**
+ * @typedef {Object} SetTmpKwargs
+ * @property {boolean} [round=true]
```
Comment (szaldivar) [inline src/models/set_tmp.js:12]: maybe we need descriptions for these properties?

File: src/models/set_tmp.js [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+) {
+  const defaults_kwargs = {
+    calculate_ce: false,
+    round: true,
+  };
+
+  let joint_kwargs = Object.assign(defaults_kwargs, kwargs);
+
+  if (units === "IP") {
+    if (body_surface_area === 1.8258) {
```
Comment (szaldivar) [inline src/models/set_tmp.js:77]: don't set the default value in the parameters list, and detect if they are undefined

File: src/models/set_tmp.js [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+      body_surface_area = 19.65;
+    }
+    if (p_atm === 101325) {
```
Comment (szaldivar) [inline src/models/set_tmp.js:80]: same as previous here

File: src/models/set_tmp.js [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+  ).set;
+
+  if (units === "IP") {
+    const convertedSet = units_converter(
+      {
+        tmp: set_tmp,
+      },
+      "SI",
+    );
+    set_tmp = convertedSet.tmp;
```
Comment (szaldivar) [inline src/models/set_tmp.js:122]: Just as a heads up, you can use this syntax to deconstruct values from the units_converter:
```javascript
({tmp: set_tmp} = units_converter({tmp: set_tmp}, "SI"))
```
that will do the same as you do inside this if but you don't have to change it. Just wanted to mentioned it

File: src/models/set_tmp.js [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+  }
+
+  if (limit_inputs) {
+    let warnings = check_standard_compliance("ashrae", {
```
Comment (szaldivar) [inline src/models/set_tmp.js:126]: are you sure the `check_standard_compliance` and `check_standard_compliance_array` check the same limits?

File: src/models/set_tmp.js [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    trArray,
+    vArray,
+    rhArray,
+    metArray,
+    cloArray,
+    wmeArray,
+    bodySurfaceArray,
+    pAtmArray,
+    bodyPositionArray,
+    [90],
```
Comment (szaldivar) [inline src/models/set_tmp.js:258]: this breaks if the arrays are not of length 1

Comment (szaldivar) [reply src/models/set_tmp.js:258]: that value doesn't do anything in `two_nodes` can you check the implementation of two_nodes pls

File: src/models/set_tmp.js [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+      clo: cloValid,
+    } = check_standard_compliance_array("ashrae", {
+      tdb: tdbArray,
+      tr: trArray,
+      v: vArray,
+      met: metArray,
+      clo: cloArray,
+    });
+
+    const allValid = !(
```
Comment (szaldivar) [inline src/models/set_tmp.js:287]: omit all this and keep just keep the for loop

**Review by szaldivar — CHANGES_REQUESTED** [review]

File: src/models/set_tmp.js (hunk 5 above)
Comment (YuchenMu) [reply src/models/set_tmp.js:144]: In the Python version, I can only tell it is checking "ashrae". I did not see other examples in the documentation of Python version

File: src/models/set_tmp.js [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  if (wme === undefined) {
+    wme = 0;
+  }
+
+  if (body_position === undefined) {
+    body_position = "standing";
+  }
+
+  if (body_surface_area === undefined) {
+    body_surface_area = 1.8258;
```
Comment (szaldivar) [inline src/models/set_tmp.js:84]: assign this value depending on the unit system so you avoid doing it below

Comment (szaldivar) [reply src/models/set_tmp.js:84]:
```javascript
body_surface_area = units === "SI" ? 1.8258 : 19.65;
```

File: src/models/set_tmp.js [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+  if (units === undefined) {
+    units = "SI";
+  }
+
+  if (limit_inputs === undefined) {
+    limit_inputs = true;
+  }
+
+  if (units === "IP") {
+    if (body_surface_area === 1.8258) {
```
Comment (szaldivar) [inline src/models/set_tmp.js:100]: remove this and do it in the undefined checks above

File: src/models/set_tmp.js (hunk 3 above)
Comment (szaldivar) [reply src/models/set_tmp.js:103]: remove this and do the same as my suggestion for `body_surface_area`

File: src/models/set_tmp.js [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    } = check_standard_compliance_array("ashrae", {
+      tdb: [tdb],
+      tr: [tr],
+      v: [v],
+      met: [met],
+      clo: [clo],
+    });
+
+    if (
+      isNaN(tdb_valid) ||
```
Comment (szaldivar) [inline src/models/set_tmp.js:158]: these are arrays now, you need to index into them, do `isNaN(tdb_valid[0])`

File: src/models/set_tmp.js [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    bodySurfaceArray = tdbArray.map((_) => 1.8258);
+  }
+
+  let p_atm = 101325;
+  if (pAtmArray === undefined) {
+    pAtmArray = tdbArray.map((_) => 101325);
+  }
+
+  if (units === "IP") {
+    if (body_surface_area === 1.8258) {
```
Comment (szaldivar) [inline src/models/set_tmp.js:262]: same comment as the other function

File: src/models/two_nodes.js [hunk 12]
```diff
@@ -133,8 +140,49 @@ export function two_nodes(
   };
 
   let joint_kwargs = Object.assign(defaults_kwargs, kwargs);
+
+  if (wme === undefined) {
+    wme = 0;
```
Comment (szaldivar) [inline src/models/two_nodes.js:145]: keep the default values in the parameters here.
Maybe you did not understand why in the other function it was better to leave them to undefined
In the other function you needed two different default values depending in the unit system, so it is better to leave it to undefined and then assign them based in the unit system like
```javascript
some_var = units === "SI" ? si_default_value : ip_default_value
```
In here there is no point in doing it this way, it is way better to set in it in the parameters

**Review by szaldivar — CHANGES_REQUESTED** [review]

File: src/models/set_tmp.js [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+  if (wmeArray === undefined) {
+    wmeArray = tdbArray.map((_) => 0);
+  }
+
+  if (bodyPositionArray === undefined) {
+    bodyPositionArray = tdbArray.map((_) => "standing");
+  }
+
+  if (units === undefined) {
+    units = "SI";
```
Comment (szaldivar) [inline src/models/set_tmp.js:233]: move to parameters

File: src/models/set_tmp.js [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  if (limit_inputs === undefined) {
+    limit_inputs = true;
```
Comment (szaldivar) [inline src/models/set_tmp.js:237]: move to parameters

File: src/models/set_tmp.js [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+
+  if (bodySurfaceArray === undefined) {
+    bodySurfaceArray = tdbArray.map((_) => 1.8258);
+  }
+
+  if (pAtmArray === undefined) {
+    pAtmArray = tdbArray.map((_) => 101325);
+  }
+
+  let body_surface_area = units === "SI" ? 1.8258 : 19.65;
```
Comment (szaldivar) [inline src/models/set_tmp.js:248]: you need this to be an array

File: src/models/set_tmp.js [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+  let p_atm = units === "SI" ? 101325 : 1;
```
Comment (szaldivar) [inline src/models/set_tmp.js:249]: needs to be an array right?

File: src/models/set_tmp.js [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+  if (units === undefined) {
+    units = "SI";
+  }
+
+  if (limit_inputs === undefined) {
+    limit_inputs = true;
+  }
+
+  if (bodySurfaceArray === undefined) {
+    bodySurfaceArray = tdbArray.map((_) => 1.8258);
```
Comment (szaldivar) [inline src/models/set_tmp.js:241]: what if units are `"IP"` ?

File: src/models/set_tmp.js [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  if (pAtmArray === undefined) {
+    pAtmArray = tdbArray.map((_) => 101325);
```
Comment (szaldivar) [inline src/models/set_tmp.js:245]: what if units are `"IP"` ?

**Review by szaldivar — CHANGES_REQUESTED** [review]

File: src/models/set_tmp.js [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+ * @see {@link set_tmp_array} for a version that supports arrays
+ *
+ * @param {number} tdb Dry bulb air temperature, default in [°C] in [°F] if `units` = 'IP'.
+ * @param {number} tr Mean radiant temperature, default in [°C]
+ * @param {number} v Air speed, default in [m/s]
+ * @param {number} rh Relative humidity, [%].
+ * @param {number} met Metabolic rate, [W/(m2)]
+ * @param {number} clo Clothing insulation, [clo]
+ * @param {number} [wme=0] External work, [W/(m2)] default 0
+ * @param {number} body_surface_area Body surface area, default value 1.8258 [m2] in [ft2] if units = ‘IP’
```
Comment (szaldivar) [inline src/models/set_tmp.js:42]:
```suggestion
 * @param {number} [body_surface_area] Body surface area, default value 1.8258 [m2] in [ft2] if units = ‘IP’
```

File: src/models/set_tmp.js [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+ * @param {number} p_atm Atmospheric pressure, default value 101325 [Pa] in [atm] if units = ‘IP’
```
Comment (szaldivar) [inline src/models/set_tmp.js:43]:
```suggestion
 * @param {number} [p_atm] Atmospheric pressure, default value 101325 [Pa] in [atm] if units = ‘IP’
```

File: src/models/set_tmp.js [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+  }
+
+  if (limit_inputs) {
+    const {
+      tdb: tdb_valid,
+      tr: tr_valid,
+      v: v_valid,
+      met: met_valid,
+      clo: clo_valid,
+    } = check_standard_compliance_array("ashrae", {
```
Comment (szaldivar) [inline src/models/set_tmp.js:122]:
```suggestion
    } = check_standard_compliance_array("ASHRAE", {
```

File: src/models/set_tmp.js [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+  }
+
+  if (limit_inputs) {
+    const {
+      tdb: tdbValid,
+      tr: trValid,
+      v: vValid,
+      met: metValid,
+      clo: cloValid,
+    } = check_standard_compliance_array("ashrae", {
```
Comment (szaldivar) [inline src/models/set_tmp.js:269]:
```suggestion
    } = check_standard_compliance_array("ASHRAE", {
```

**Review by szaldivar — APPROVED** [review]

**Pull request merged**